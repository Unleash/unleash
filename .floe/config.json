{
  "$schema": "https://unpkg.com/@floe/config@0.1.0-beta.13/schema.json",
  "reviews": {
    "maxFileEvaluations": 33,
    "maxDiffEvaluations": 20
  },
  "rulesets": {
    "docs": {
      "include": ["**/*.md", "**/*.mdx"],
      "rules": {
        "spelling-and-grammar": "warn"
      }
    }
  }
}
