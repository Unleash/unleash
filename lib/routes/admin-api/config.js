'use strict';

const Controller = require('../controller');

class ConfigController extends Controller {
    constructor(config) {
        super(config);
        this.uiConfig = { ...config.ui, version: config.version };

        this.get('/', this.getUIConfig);
    }

    async getUIConfig(req, res) {
        const config = this.uiConfig;
        res.json(config);
    }
}

module.exports = ConfigController;
