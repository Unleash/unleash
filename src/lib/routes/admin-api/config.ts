import { Response } from 'express';
import { AuthedRequest } from '../../types/core';
import { IUnleashServices } from '../../types/services';
import { IAuthType, IUnleashConfig } from '../../types/option';
import version from '../../util/version';
import Controller from '../controller';
import VersionService from '../../services/version-service';
import SettingService from '../../services/setting-service';
import {
    simpleAuthSettingsKey,
    SimpleAuthSettings,
} from '../../types/settings/simple-auth-settings';
import { ADMIN, NONE } from '../../types/permissions';
import { createResponseSchema } from '../../openapi/util/create-response-schema';
import {
    uiConfigSchema,
    UiConfigSchema,
} from '../../openapi/spec/ui-config-schema';
import { OpenApiService } from '../../services/openapi-service';
import { EmailService } from '../../services/email-service';
import { emptyResponse } from '../../openapi/util/standard-responses';
import { IAuthRequest } from '../unleash-types';
import { extractUsername } from '../../util/extract-user';
import NotFoundError from '../../error/notfound-error';
import { SetUiConfigSchema } from '../../openapi/spec/set-ui-config-schema';
import { createRequestSchema } from '../../openapi/util/create-request-schema';

class ConfigController extends Controller {
    private versionService: VersionService;

    private settingService: SettingService;

    private emailService: EmailService;

    private readonly openApiService: OpenApiService;

    constructor(
        config: IUnleashConfig,
        {
            versionService,
            settingService,
            emailService,
            openApiService,
        }: Pick<
            IUnleashServices,
            | 'versionService'
            | 'settingService'
            | 'emailService'
            | 'openApiService'
        >,
    ) {
        super(config);
        this.versionService = versionService;
        this.settingService = settingService;
        this.emailService = emailService;
        this.openApiService = openApiService;

        this.route({
            method: 'get',
            path: '',
            handler: this.getUiConfig,
            permission: NONE,
            middleware: [
                openApiService.validPath({
                    tags: ['Admin UI'],
                    operationId: 'getUiConfig',
                    responses: {
                        200: createResponseSchema('uiConfigSchema'),
                    },
                }),
            ],
        });

        this.route({
            method: 'post',
            path: '',
            handler: this.setUiConfig,
            permission: ADMIN,
            middleware: [
                openApiService.validPath({
                    tags: ['Admin UI'],
                    operationId: 'setUiConfig',
                    requestBody: createRequestSchema('setUiConfigSchema'),
                    responses: { 200: emptyResponse },
                }),
            ],
        });
    }

    async getUiConfig(
        req: AuthedRequest,
        res: Response<UiConfigSchema>,
    ): Promise<void> {
        const [frontendSettings, simpleAuthSettings] = await Promise.all([
            this.settingService.getFrontendSettings(),
            this.settingService.get<SimpleAuthSettings>(simpleAuthSettingsKey),
        ]);

        const disablePasswordAuth =
            simpleAuthSettings?.disabled ||
            this.config.authentication.type == IAuthType.NONE;

        const expFlags = this.config.flagResolver.getAll({
            email: req.user.email,
        });
        const flags = { ...this.config.ui.flags, ...expFlags };

        const response: UiConfigSchema = {
            ...this.config.ui,
            flags,
            version,
            emailEnabled: this.emailService.isEnabled(),
            unleashUrl: this.config.server.unleashUrl,
            baseUriPath: this.config.server.baseUriPath,
            authenticationType: this.config.authentication?.type,
            segmentValuesLimit: this.config.segmentValuesLimit,
            strategySegmentsLimit: this.config.strategySegmentsLimit,
            frontendApiOrigins: frontendSettings.frontendApiOrigins,
            versionInfo: this.versionService.getVersionInfo(),
            disablePasswordAuth,
        };

        this.openApiService.respondWithValidation(
            200,
            res,
            uiConfigSchema.$id,
            response,
        );
    }

    async setUiConfig(
        req: IAuthRequest<void, void, SetUiConfigSchema>,
        res: Response<string>,
    ): Promise<void> {
        if (req.body.frontendSettings) {
            await this.settingService.setFrontendSettings(
                req.body.frontendSettings,
                extractUsername(req),
            );
            res.sendStatus(204);
            return;
        }

        throw new NotFoundError();
    }
}

export default ConfigController;
