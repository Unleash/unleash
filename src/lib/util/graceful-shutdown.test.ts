import {
    executeShutdownHooks,
    registerGracefulShutdownHook,
} from './graceful-shutdown';
import { createTestConfig } from '../../test/config/test-config';

describe('Graceful shutdown hooks', () => {
    test('Shutdown hooks can be registered', async () => {
        const config = createTestConfig({});
        let wasCalled = false;
        registerGracefulShutdownHook(
            config.getLogger('shutdown-hook'),
            'test-shutdown',
            async () => {
                wasCalled = true;
            },
        );
        await executeShutdownHooks(config.getLogger('hook-executer'));
        expect(wasCalled).toBeTruthy();
    });
    test('Registering hook with same name as existing hook, should throw error', async () => {
        const config = createTestConfig({});
        let wasCalled = false;
        registerGracefulShutdownHook(
            config.getLogger('shutdown-hook'),
            'test-shutdown',
            async () => {
                wasCalled = true;
            },
        );
        expect(() =>
            registerGracefulShutdownHook(
                config.getLogger('shutdown-hook'),
                'test-shutdown',
                async () => {
                    wasCalled = true;
                },
            ),
        ).toThrowErrorMatchingSnapshot('duplicate-shutdown-hook');
    });
    test('Can register multiple hooks with different names', async () => {
        const config = createTestConfig({});
        let wasCalled = false;
        let secondWasCalled = false;
        registerGracefulShutdownHook(
            config.getLogger('shutdown-hook'),
            'hook-1',
            async () => {
                wasCalled = true;
            },
        );
        registerGracefulShutdownHook(
            config.getLogger('shutdown-hook'),
            'hook-2',
            async () => {
                secondWasCalled = true;
            },
        );
        await executeShutdownHooks(config.getLogger('graceful-shutdown'));
        expect(wasCalled).toEqual(true);
        expect(secondWasCalled).toEqual(true);
    });

    test('Shutdown will wait on all hooks', async () => {
        const config = createTestConfig({});
        let wasCalled = false;
        let secondWasCalled = false;
        registerGracefulShutdownHook(
            config.getLogger('shutdown-hook'),
            'hook-1',
            async () => {
                wasCalled = true;
            },
        );
        registerGracefulShutdownHook(
            config.getLogger('shutdown-hook'),
            'hook-2',
            async () => {
                return new Promise((resolve) =>
                    setTimeout(() => {
                        secondWasCalled = true;
                        resolve();
                    }, 500),
                );
            },
        );
        await executeShutdownHooks(config.getLogger('graceful-shutdown'));
        expect(wasCalled).toEqual(true);
        expect(secondWasCalled).toEqual(true);
    });
    test('Calling executeShutdownHooks should empty the registered shutdown hooks', async () => {
        const config = createTestConfig({});
        let wasCalled = false;
        registerGracefulShutdownHook(
            config.getLogger('shutdown-hook'),
            'hook-1',
            async () => {
                wasCalled = true;
            },
        );
        await executeShutdownHooks(config.getLogger('graceful-shutdown'));
        expect(() =>
            registerGracefulShutdownHook(
                config.getLogger('shutdown-hook'),
                'hook-1',
                async () => {
                    wasCalled = true;
                },
            ),
        ).not.toThrow();
    });
});
