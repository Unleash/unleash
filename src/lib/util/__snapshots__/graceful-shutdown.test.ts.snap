// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Graceful shutdown hooks Registering hook with same name as existing hook, should throw error: duplicate-shutdown-hook 1`] = `"test-shutdown already registered for graceful shutdown"`;
